### PR TITLE
Adding limits to container start operations

### PR DESCRIPTION
**Pull request by alshabib**

### CODE REVIEW

File: containerz/containerz.proto [hunk 1]
```diff
@@ -417,6 +417,30 @@ message StartContainerRequest {
   RunAs run_as = 11;
 
   map<string, string> labels = 12;
+  
+  // Limits defines restrictions or limits upon this container.
+  message Limits {
+    // max_cpu specifies the maximum amount of cpu this container can use.
+    // For example a value of 0.5 means this container can use a maximum of
+    // half a cpu. This setting does not imply that the container is pinned
```
Comment (robshakir) [inline containerz/containerz.proto:425]: Is this half a CPU or half a core? (I assume that this means 50% of a core that is exposed to the kernel -- which I think is referred to as a CPU in some commands, e.g., `top(1)`.) k8s uses the definition:

> 1 Hyperthread on a bare-metal Intel processor with Hyperthreading [ref](https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/)

I presume that this is going to some underlying cgroups API eventually, so the definition could come from there.

File: containerz/containerz.proto [hunk 2]
```diff
@@ -417,6 +417,30 @@ message StartContainerRequest {
...
+    // to a cpu but rather only the maximum amount of cpu it can use.
+    double max_cpu = 1;
+
+    // soft_mem defines a soft memory limit. If the system is not memory
+    // constrained this limit has no effect. If it is memory constrained
+    // this container may be oom killed if it exceed this limit. This value
+    // is effectively a soft memory limit. This limit should be smaller than
+    //  max_mem otherwise it will have no effect.
+    int64 soft_mem =2;
```
Comment (robshakir) [inline containerz/containerz.proto:434]: What are the units here? Generally, I think best practice is to have the units in the name -- e.g., `soft_mem_bytes`.

File: containerz/containerz.proto [hunk 3]
```diff
@@ -417,6 +417,30 @@ message StartContainerRequest {
...
+
+    // hard_mem is a hard limit. If the container exceed this number the 
+    // container runtime can take aggressive action such as terminate (oom kill)
+    // the container.
+    int64 hard_mem = 3;
```
Comment (robshakir) [inline containerz/containerz.proto:439]: Same question above re: units.

**Review by robshakir — CHANGES_REQUESTED** [review]

File: containerz/containerz.proto (hunk 1 above)
Comment (alshabib) [reply containerz/containerz.proto:425]: Hmm googd point - I think we should define this as a CPU and the specific implementation for the container runtime can translate accordingly.

File: containerz/containerz.proto (hunk 2 above)
Comment (alshabib) [reply containerz/containerz.proto:434]: ah yes - thanks!

File: containerz/containerz.proto (hunk 3 above)
Comment (alshabib) [reply containerz/containerz.proto:439]: Thanks!

File: containerz/containerz.proto (hunk 1 above)
Comment (robshakir) [reply containerz/containerz.proto:425]: I think we can just say "an OS-visible CPU (which may map to a hyperthread on an Intel processor with hyperthreading, or a CPU core".

**Review by robshakir — APPROVED** [review]

**Pull request merged**